lib/multipass_ex.ex

```diff
@@ -1,79 +1,34 @@
-      formatted_data =
-        data
-        |> Jason.encode!()
-        |> pad(@aes128_block_size)
+  If integrating with a service using the Rails gem care must be taken to ensure
+  `site_key` and `api_key` match between the services.
 
-      encoded_data =
-        :aes_cbc128
-        |> :crypto.block_encrypt(generate_key(site_key, api_key), @iv, formatted_data)
-        |> Base.url_encode64()
+  ## Example
 
-      {:ok, encoded_data}
-    rescue
-      err -> {:error, err}
-    end
-  end
+      iex> MultipassEx.encode(%{example: "data"}, "my_site_key", "my_super_secret_key_shh")
+      {:ok, "HFtLULqnAfVZexe46_T5KHJcgSRwAtf45zPwJ5g361w"}
 
-  @doc """
-  Decodes a given multipass into a map with a given site_key and api_key
+      iex> MultipassEx.decode("HFtLULqnAfVZexe46_T5KHJcgSRwAtf45zPwJ5g361w", "my_site_key", "my_super_secret_key_shh")
+      {:ok, %{"example" => "data"}}
   """
-  @spec decode(String.t(), String.t(), String.t()) :: {:ok, map()} | {:error, String.t()} | {:error, any()}
-  def decode(data, site_key, api_key)
-      when is_binary(data) and is_binary(site_key) and is_binary(api_key) do
-    try do
-      formatted_data = Base.url_decode64!(data)
-
-      decoded_data =
-        :aes_cbc128
-        |> :crypto.block_decrypt(generate_key(site_key, api_key), @iv, formatted_data)
-        |> unpad()
-        |> Jason.decode!()
-
-      {:ok, decoded_data}
-    rescue
-      e in ArgumentError -> {:error, e.message}
-      err -> {:error, err}
-    end
-  end
 
-  @spec generate_key(String.t(), String.t()) :: binary()
-  defp generate_key(password, salt) do
-    :crypto.hash(:sha, salt <> password) |> truncate_secret()
-  end
+  alias MultipassEx.{Crypto, CryptoError, DecodingError, EncodingError, JSONDecodingError}
 
-  @spec truncate_secret(String.t()) :: String.t()
-  defp truncate_secret(secret) do
-    <<key::size(@encryption_bits), _rest::binary>> = secret
-    <<key::size(@encryption_bits)>>
-  end
-
-  @doc false
-  @spec pad(String.t(), pos_integer()) :: String.t()
-  def pad(data, block_size) when is_binary(data) and is_integer(block_size) and block_size >= 0 do
-    to_add = block_size - rem(byte_size(data), block_size)
-    data <> :binary.copy(<<to_add>>, to_add)
-  end
+  @doc """
+  Encodes a map resulting in a Multipass compliant string.
+  """
+  @spec encode(map(), String.t(), String.t()) ::
+          {:ok, String.t()}
+          | {:error, EncodingError.t()}
+  defdelegate encode(data, site_key, api_key), to: Crypto, as: :encrypt
 
-  @doc false
-  @spec unpad(binary()) :: binary()
-  def unpad(data) when is_binary(data) do
-    to_remove = :binary.last(data)
-    :binary.part(data, 0, byte_size(data) - to_remove)
-  end
+  @doc """
+  Decodes a Multipass string into a map.
+  """
+  @spec decode(String.t(), String.t(), String.t()) ::
+          {:ok, map()} | {:error, DecodingError.t() | CryptoError.t() | JSONDecodingError.t()}
+  defdelegate decode(data, site_key, api_key), to: Crypto, as: :decrypt
```

Comment:
As a client of this API, do you see some pattern matching differently on a `DecodingError` vs a `CryptoError` vs a `JSONDecodingError`? If not, would a single `DecodingError` with a reason atom (and maybe a message) work? That might simplify the error API the client has to handle.